### ISSUE TITLE
Exclude completely "target" from Eclipse in Checkstyle

### ISSUE THREAD

**Issue by cristcost** [m2e-code-quality/m2e-code-quality#39]
Hello,

My Maven project generate classes inside the folder "target/generated/src/main/java".

The problem is that the when I import it into eclipse, eclipse-cs get configured to validate also the java classes inside that folder as they match the pattern "src/main/java/.*\/.*\.java"

To be able to exclude the generated files, I need to change configuration in eclipse either by adding an exclude rule for the pattern "^target" or change the actual rule to "^src/main/java/.*\/.*\.java", so that only paths *starting* with that get checked.

I've tried to work on <includes> or <excludes> parameters of the checkstyle plugin in the pom.xml, but if I set
```
<excludes>^target</excludes>
```
in Eclipse I get added an exclusion rule of 
```
src/main/java.*^target
```
and if I set a rule like
```
<includes>^src/main/java/.*\/.*\.java</includes>
```
in Eclipse I get the main pattern changed to 
```
src/main/java/^src/main/java/\..*\/\..*\\.java
```

I think that maybe m2e-code-quality needs to handle this, either by 
* configuring eclipse automatically with a pattern starting with "^", 
* or adding an exclusion for the "^target", 
* or by allowing to achieve this result from configuration in pom.xml

**Comment by ArloL** [comment]
Hi @cristcost,

I am trying to cleanup the issues of the project. Is this still a problem?

**Comment by sellersj** [comment]
This is a problem for me.

It's possible to work around it by creating a suppression file and linking it to the checkstyle file.

However, I don't know why anyone would want to check the code quality for anything that's in ${project.build.directory} e.g. target.

Can the plugin be changed to exclude the derived (generated) files by default?

Your comments in https://github.com/m2e-code-quality/m2e-code-quality/issues/80 are very helpful.

**Comment by sellersj** [comment]
I've found a work around for checkstyle. I've just excluded anything it from checking anything in the target directory and it seems to work well enough for now.

Added config to my pom:
<excludes>${project.build.directory}/**</excludes>

**Comment by ArloL** [comment]
Are there checkstyle issues when you run mvn? Because there should be no difference between eclipse-cs and maven-checkstyle-plugin behavior.

If there is, could you create a PR with checkstyle-8 in https://github.com/m2e-code-quality/m2e-code-quality/tree/develop/test-projects that demonstrates it? Then I can have a look and try fixing the issue.

**Comment by ArloL** [comment]
Hi @sellersj, do you still have problems?

**Comment by mekazu** [comment]
Hi @ArloL I wouldn't know how to to raise a PR for this but I completely agree @sellersj 's first comment and in answer to your question - there is a difference between running mvn and maven-checkstyle-plugin:

Running checkstyles with mvn doesn't cause errors because the default [includes](http://maven.apache.org/archives/maven-1.x/plugins/checkstyle/properties.html) configuration is relative to the src dir i.e. 'src/main/java' of the project, so it already excludes 'target'. This plugin shows up checkstyle errors for all java files in the eclipse classpath of the project.

The fix would be for this plugin to only show checkstyle errors for java files that are in the configured 'includes' locations excluding the 'excludes' location to make it consistent with the maven plugin.

http://maven.apache.org/archives/maven-1.x/plugins/checkstyle/properties.html

**Comment by ArloL** [comment]
Hi @mekazu, are you sure you are referring to the latest maven-checkstyle-plugin configuration? The documentation I find includes all java files with `**\/*.java`. See https://maven.apache.org/plugins/maven-checkstyle-plugin/checkstyle-mojo.html#includes

**Comment by mekazu** [comment]
Hi @ArloL you're right so I suppose then the reason why the maven checkstyles plugin doesn't create any errors is because the checkstyle phase occurs before the generate phase, so when checkstyle runs there aren't any generated files for it to validate.

Therefore I'd say the solution is simply to ignore any java source files in the target directory.

**Comment by ArloL** [comment]
Please have a look at the [checkstyle-8 demo project](https://github.com/m2e-code-quality/m2e-code-quality/tree/develop/test-projects/checkstyle-8). The maven-checkstyle-plugin correctly reports problems since it includes all [sourceDirectories](https://maven.apache.org/plugins/maven-checkstyle-plugin/checkstyle-mojo.html#sourceDirectories) by default.

~~~~
[INFO] --- maven-checkstyle-plugin:2.14:check (default) @ checkstyle-8 ---
[INFO] There are 32 checkstyle errors.
[ERROR] src/main/java/cat/App.java[0] (javadoc) JavadocPackage: Missing package-info.java file.
[ERROR] src/main/java/cat/App.java[7] (regexp) RegexpSingleline: Line has trailing spaces.
[ERROR] src/main/java/cat/App.java[7:1] (design) HideUtilityClassConstructor: Utility classes should not have a public or default constructor.
[ERROR] src/main/java/cat/App.java[8:1] (blocks) LeftCurly: '{' should be on the previous line.
[ERROR] src/main/java/cat/App.java[9:5] (javadoc) JavadocMethod: Missing a Javadoc comment.
[ERROR] src/main/java/cat/App.java[9:29] (whitespace) ParenPad: '(' is followed by whitespace.
[ERROR] src/main/java/cat/App.java[9:30] (misc) FinalParameters: Parameter args should be final.
[ERROR] src/main/java/cat/App.java[9:43] (whitespace) ParenPad: ')' is preceded with whitespace.
[ERROR] src/main/java/cat/App.java[10:5] (blocks) LeftCurly: '{' should be on the previous line.
[ERROR] src/main/java/cat/App.java[11:5] (whitespace) FileTabCharacter: Line contains a tab character.
[ERROR] src/main/java/cat/App.java[11:38] (coding) MagicNumber: '12345' is a magic number.
[ERROR] src/main/java/cat/App.java[12:5] (whitespace) FileTabCharacter: Line contains a tab character.
[ERROR] src/main/java/cat/App.java[13:5] (whitespace) FileTabCharacter: Line contains a tab character.
[ERROR] src/main/java/cat/App.java[14:5] (whitespace) FileTabCharacter: Line contains a tab character.
[ERROR] src/main/java/cat/App.java[14:28] (whitespace) ParenPad: '(' is followed by whitespace.
[ERROR] src/main/java/cat/App.java[16:28] (whitespace) ParenPad: '(' is followed by whitespace.
[ERROR] target/generated/src/main/java/cat/App2.java[0] (javadoc) JavadocPackage: Missing package-info.java file.
[ERROR] target/generated/src/main/java/cat/App2.java[7] (regexp) RegexpSingleline: Line has trailing spaces.
[ERROR] target/generated/src/main/java/cat/App2.java[7:1] (design) HideUtilityClassConstructor: Utility classes should not have a public or default constructor.
[ERROR] target/generated/src/main/java/cat/App2.java[8:1] (blocks) LeftCurly: '{' should be on the previous line.
[ERROR] target/generated/src/main/java/cat/App2.java[9:5] (javadoc) JavadocMethod: Missing a Javadoc comment.
[ERROR] target/generated/src/main/java/cat/App2.java[9:29] (whitespace) ParenPad: '(' is followed by whitespace.
[ERROR] target/generated/src/main/java/cat/App2.java[9:30] (misc) FinalParameters: Parameter args should be final.
[ERROR] target/generated/src/main/java/cat/App2.java[9:43] (whitespace) ParenPad: ')' is preceded with whitespace.
[ERROR] target/generated/src/main/java/cat/App2.java[10:5] (blocks) LeftCurly: '{' should be on the previous line.
[ERROR] target/generated/src/main/java/cat/App2.java[11:5] (whitespace) FileTabCharacter: Line contains a tab character.
[ERROR] target/generated/src/main/java/cat/App2.java[11:38] (coding) MagicNumber: '12345' is a magic number.
[ERROR] target/generated/src/main/java/cat/App2.java[12:5] (whitespace) FileTabCharacter: Line contains a tab character.
[ERROR] target/generated/src/main/java/cat/App2.java[13:5] (whitespace) FileTabCharacter: Line contains a tab character.
[ERROR] target/generated/src/main/java/cat/App2.java[14:5] (whitespace) FileTabCharacter: Line contains a tab character.
[ERROR] target/generated/src/main/java/cat/App2.java[14:28] (whitespace) ParenPad: '(' is followed by whitespace.
[ERROR] target/generated/src/main/java/cat/App2.java[16:28] (whitespace) ParenPad: '(' is followed by whitespace.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
~~~~

Or am I missing something?

**Comment by ArloL** [comment]
Were you able to reproduce my findings, @mekazu?

**Comment by mekazu** [comment]
I think the project you've provided is a bad example (no offence!) because the generated sources in the target directory are usually generated by some other plugin. For example, a cxf maven plugin might be used to convert xsds to java models or a wsdl to some java interfaces. The project author has no control (or interest) over the formatting of the generated sources.

If in the above project the generated/src/main/java directory was a source folder then sure I'd expect to see checkstyle errors because it's not in the target directory and is part of the source of the project.

Basically I'm recommending for a blanket rule to avoid performing checkstyle checks of files residing in the target directory, even if they are a source directory.

**Comment by antonsmyk** [comment]
> to avoid performing checkstyle checks of files residing in the target directory, even if they are a source directory.

They are in so called "source roots" which in https://issues.apache.org/jira/browse/MCHECKSTYLE-70 since maven-checkstyle-plugin 2.13 means generated sources also. It is rather unhappy unwelcome and unjustified change. See also https://issues.apache.org/jira/browse/MCHECKSTYLE-318.

**Comment by antonsmyk** [comment]
The maven-checkstyle-plugin:check goal binds to "verify" goal by default when all sources are generated already. But it still complains on generated sources in Eclipse even when bind it to "validate" phase, and have explicit configuration for the plugin to check only files under src/main/java directory:
```
<configuration>
  <sourceDirectories>
    <sourceDirectory>src/main/java</sourceDirectory>
  </sourceDirectories>
  <testSourceDirectories>
    <testSourceDirectory>src/test/java</testSourceDirectory>
  </testSourceDirectories>
</configuration>
``` 

@ArloL, it looks like m2e totally ignores maven-checkstyle-plugin sources configuration, isn't it?

**Comment by antonsmyk** [comment]
I tried with Eclipse Mars 2 with embedded Maven 3.3.3 and with external Maven 3.3.9.

**Comment by antonsmyk** [comment]
See pull request: I have add a test case demonstrating the issue: generated sources ignored in Maven command line execution, but Eclipse still complains on them.

Screenshot:
![eclipse-checkstyle-39-test-project](https://cloud.githubusercontent.com/assets/10366539/15577233/f00b3978-235b-11e6-82a2-33b0b0c9186b.png)

**Comment by ArloL** [comment]
Thanks for the project that demonstrates the issue, @antonsmyk. That was a great help.

The problem is that eclipse-cs works with project relative regular expressions for the source directories. So what I do now is to add a `^` to the beginning of those regular expressions.

I have updated the snapshot update site: https://m2e-code-quality.github.io/m2e-code-quality/site/develop/

If you could please test and report back that everything works as expected I can update latest. Thanks for your help!

**Comment by antonsmyk** [comment]
Thanks for prompt action, @ArloL! Unfortunately, problem is still there. I've used "Checkstyle configuration plugin for M2Eclipse 1.0.0.201605281046".

I've also tried moving maven-checkstyle-plugin configuration to pluginManagement also, but no difference. 

In Project->Properties->Checkstyle->Edit I see that generated sources directory is present and activated:

![project-properties-checkstyle](https://cloud.githubusercontent.com/assets/10366539/15644668/9e5a3f74-2654-11e6-9bf4-ee9d60aac4d9.png)

![project-properties-checkstyle-edit](https://cloud.githubusercontent.com/assets/10366539/15644671/a47238da-2654-11e6-954c-6750f6680028.png)

Manually unticking "^target/generated/src/main/java.*\.java" entry helps -- Eclipse stops complaining on files in that directory. But I want this entry to never present in the first place due to maven-checkstyle-plugin configuration in project POM.

**Comment by antonsmyk** [comment]
Also tried with latest maven-checkstyle-plugin version 2.17 stated in POM (instead of 2.14) with the same result.

**Comment by ArloL** [comment]
I have updated the develop site with a new snapshot.

It seems this is a bug in m2e since it never actually resolves the `sourceDirectories` property. It just returns the default value. With some trickery I got it working: https://github.com/m2e-code-quality/m2e-code-quality/commit/3a87bd9a939c35a5d4408342d0491906769da648

**Comment by antonsmyk** [comment]
Yes, 1.0.0.201605301854 works fine! Thank you very much!

![eclipse-checkstyle-correct-list-of-patterns](https://cloud.githubusercontent.com/assets/10366539/15669018/742b4818-271e-11e6-9df2-319ef1a3e8d0.png)

**Comment by ArloL** [comment]
Can you confirm that it works, @mekazu?

**Comment by mekazu** [comment]
I'm happy to take @antonsmyk 's word for it.

**Comment by ArloL** [comment]
If you don't have the time to test this, that is absolutely understandable. I just want to make sure that I don't break anything. My confidence in releasing increases with every successful test by other people. So if you get a chance to test the snapshot, that would be great! Thanks.

**Comment by ArloL** [comment]
Did anybody else have time to test this?

**Comment by ArloL** [comment]
This has been released now. Please let me know if there are any issues.
